tensorflow/python/kernel_tests/sparse_tensor_dense_matmul_op_base.py

```diff
@@ -67,9 +67,10 @@ def _testMatmul(self,
 
       sp_x_value = sparse_tensor.SparseTensorValue(
           indices=x_indices, values=x_values, dense_shape=x_shape)
-
+      gpus = config.list_physical_devices('GPU')
       if (self.__class__.__name__ == "SparseTensorDenseMatmulDeterministicTest"
-          and x.dtype in self._getGpuDeterminismUnimplementedTypes()):
+          and x.dtype in self._getGpuDeterminismUnimplementedTypes() and
+          len(gpus) > 0):
```

Comment:
This highlights a shortcoming in the test-organization pattern that I created and that we have been using. When `SparseTensorDenseMatMulDeterministicTest::main()` runs, the regular op tests inherited from `SparseTensorDenseMatMulTestBase` will run even when a GPU is not available. This is wasteful.

Rather than adding this condition here, I wonder if it makes sense to run `SparseTensorDenseMatMulDeterministicTest::main()` only if GPUs are available. This will improve encapsulation of the deterministic testing and also reduce the amount of wasted compute. Then none of the base tests will be run at all as part of deterministic testing when there are no GPUs available.

Can you see a more elegant/proper way of doing this than what I suggest above?

Reply:
While GPU is present, all tests in `SparseTensorDenseMatMulTestBase`  still get run twice even if 
```
if len(config.list_physical_devices('GPU')) > 0:
  test.main()
```
is applied. Thinking about changing inheritance pattern from `SparseTensorDenseMatMulTestBase` such that `SparseTensorDenseMatMulDeterministicTest` only run its own test.

Reply:
> While GPU is present, all tests in SparseTensorDenseMatMulTestBase still get run twice even if
> ``` 
> if len(config.list_physical_devices('GPU')) > 0:
>   test.main()
> ```
> Thinking about changing inheritance pattern from SparseTensorDenseMatMulTestBase such that SparseTensorDenseMatMulDeterministicTest only runs its own test.

Correct. I made the pattern like that way intentionally. The base op tests are run with both with `TF_DETERMINISTIC_OPS` set to `'1'` (under `SparseTensorDenseMatMulDeterministicTest`) and with `TF_DETERMINISTIC_OPS` unset/default (under `SparseTensorDenseMatMulTest`). The intention is to test both the deterministic and nondeterministic code paths for arithmetic correctness, using one set of tests (in `SparseTensorDenseMatMulTestBase`) and, ideally with only `SparseTensorDenseMatMulDeterministicTest` knowing anything about determinism.

Reply:
Ok, I will just keep the inheritance pattern.